packages/Search/src/details/DTK_DetailsTreeConstruction_decl.hpp

```diff
@@ -95,6 +101,184 @@ struct TreeConstruction
+        Kokkos::RangePolicy<ExecutionSpace>( 0, n ), KOKKOS_LAMBDA( int i ) {
+            Point xyz;
+            double a, b;
+            centroid( primitives( i ), xyz );
+            // scale coordinates with respect to bounding box of the scene
+            for ( int d = 0; d < 3; ++d )
+            {
+                a = scene_bounding_box.minCorner()[d];
+                b = scene_bounding_box.maxCorner()[d];
+                xyz[d] = ( a != b ? ( xyz[d] - a ) / ( b - a ) : 0 );
+            }
+            morton_codes( i ) = morton3D( xyz[0], xyz[1], xyz[2] );
+        } );
+    Kokkos::fence();
+}
+
+template <typename Primitives, typename MortonCodes>
+inline void assignMortonCodesDispatch( PointTag, Primitives primitives,
+                                       MortonCodes morton_codes,
+                                       Box const &scene_bounding_box )
+{
+    using ExecutionSpace =
+        typename decltype( primitives )::traits::execution_space;
+    auto const n = morton_codes.extent( 0 );
+    Kokkos::parallel_for(
+        DTK_MARK_REGION( "assign_morton_codes" ),
+        Kokkos::RangePolicy<ExecutionSpace>( 0, n ), KOKKOS_LAMBDA( int i ) {
+            Point xyz = primitives( i );
+            double a, b;
+            // scale coordinates with respect to bounding box of the scene
+            for ( int d = 0; d < 3; ++d )
+            {
+                a = scene_bounding_box.minCorner()[d];
+                b = scene_bounding_box.maxCorner()[d];
+                xyz[d] = ( a != b ? ( xyz[d] - a ) / ( b - a ) : 0 );
+            }
+            morton_codes( i ) = morton3D( xyz[0], xyz[1], xyz[2] );
+        } );
+    Kokkos::fence();
+}
+
+template <typename DeviceType>
+template <typename ConstViewType>
+inline void TreeConstruction<DeviceType>::assignMortonCodes(
+    ConstViewType primitives,
+    Kokkos::View<unsigned int *, DeviceType> morton_codes,
+    Box const &scene_bounding_box )
+{
+    using Tag = typename Tag<typename decltype(
+        primitives )::traits::non_const_value_type>::type;
+    assignMortonCodesDispatch( Tag{}, primitives, morton_codes,
+                               scene_bounding_box );
+}
+
+template <typename Primitives, typename Indices, typename Nodes>
+inline void initializeLeafNodesDispatch( BoxTag, Primitives primitives,
+                                         Indices permutation_indices,
+                                         Nodes leaf_nodes )
+{
+    using ExecutionSpace =
+        typename decltype( primitives )::traits::execution_space;
+    auto const n = leaf_nodes.extent( 0 );
+    DTK_REQUIRE( permutation_indices.extent( 0 ) == n );
+    DTK_REQUIRE( primitives.extent( 0 ) == n );
+    static_assert( sizeof( typename decltype(
+                       permutation_indices )::value_type ) == sizeof( Node * ),
+                   "Encoding leaf index in pointer to child is not safe if the "
+                   "index and pointer types do not have the same size" );
```

Comment:
TODO move assertions to `initializeLeafNodes()` to avoid duplication with tag dispatching